### ISSUE TITLE
Embedding generation writes empty .pth.gz when a shard spans multiple batches (batch_size < num_samples_per_shard)

### ISSUE THREAD

**Issue by geniusxxx** [apple/ml-mobileclip-dr#3]
### Summary
When running embedding generation on WebDataset shards whose sample count exceeds the configured batch_size, many samples in the output .tar contain a tiny .pth.gz (~1 KB) with missing image_emb/text_emb tensors. This does not happen when the entire shard fits in a single batch.
### Environment
- Ray: 2.22.0 (pip)
- Python: 3.10.18 (conda env)
- PyTorch / CUDA: 2.5.1+cu121 / 12.1
- torchvision: 0.20.1+cu121
- open_clip_torch: 2.31.0
- GPUs: 3× NVIDIA H200 NVL on a single server (8 GPUs available)
- CUDA Toolkit: 12.1.66 (nvcc -V)
- OS: Linux ubuntu 5.4.0-216-generic x86_64; ubuntu20.04
### Observed behavior
Many samples inside the generated .tar have a .pth.gz around 1 KB. Loading them shows missing or empty image_emb/text_emb: tensor image_emb[0]=0, text_emb[0]=0.
### Expected behavior
Regardless of how many batches the shard is split into, every sample should contain valid image_emb and text_emb tensors.
### Cause analysis

- This seems possibly related to cross-batch index alignment around `class EmbeddingModel.__call__` in [`scripts/gen.py`](https://github.com/apple/ml-mobileclip-dr/blob/418551d5a8efef8fb2ce829fdb1d3f1e73a357b9/scripts/gen.py#L477).

- It looks like embeddings are appended as [B, D] per batch into global lists: 

  image_emb_all += [image_embedding] where image_embedding is [B, D]
  text_emb_all += [text_embedding] where text_embedding is [B, D]

- In the final “write-back” loop, code slices using the global sample index tensor_index:
  ``` python
  image_embedding = torch.cat(
    [image_emb_all[j][tensor_index:tensor_index+1] for j in range(num_img_samples)],
    dim=0
  )
  ```
- However, image_emb_all[j] here corresponds only to the first batch’s [B1, D] for augmentation j. When tensor_index ≥ B1 (i.e., samples from later batches), the slice is empty, leading to empty concatenations and tiny .pth.gz files. The same misalignment applies to text_emb_all.
- This is just a lead; I’d appreciate maintainers’ confirmation or correction.

### Proposed fix

- Accumulate embeddings across batches indexed by augmentation index (for images) and caption index (for text). After all batches are processed, concatenate along dim=0 to build global [N, D] tensors per index; then use tensor_index to slice safely.
- Example:
  ``` python
  # Inside EmbeddingModel.__call__ (initialization)
  image_emb_per_aug = None  # List[List[Tensor]]; length == num_img_samples
  text_emb_per_cap = None   # List[List[Tensor]]; grows to max_text_samples
  
  # Per-batch, for each augmentation
  if image_emb_per_aug is None:
      image_emb_per_aug = [[] for _ in range(num_img_samples)]
  image_emb_per_aug[img_idx].append(image_embedding.cpu())
  
  # Per-batch, for each caption position
  if text_emb_per_cap is None:
      text_emb_per_cap = [[] for _ in range(max_text_samples)]
  elif len(text_emb_per_cap) < max_text_samples:
      text_emb_per_cap.extend([[] for _ in range(max_text_samples - len(text_emb_per_cap))])
  text_emb_per_cap[cap_idx].append(text_embedding.cpu())
  
  # After all batches, concatenate to global ordering per index
  image_emb_all = [torch.cat(xs, dim=0) for xs in image_emb_per_aug] if image_emb_per_aug else []
  text_emb_all  = [torch.cat(xs, dim=0) for xs in text_emb_per_cap]  if text_emb_per_cap  else []
  
  # Then the existing write-back loop slices by tensor_index as before
  ```
### Discussion
The above may not cover all cases, and there might be better fixes or intended behavior I missed. Could you help confirm whether this reasoning makes sense and suggest the preferred approach?
